# Project 1: Build Decision Tree(DV-"Survived",IDV-"Age,Gender and Fare") and Prediction

In [1]:
# Importing packages
import numpy as np
import pandas as pd
from sklearn import tree
from sklearn import preprocessing

#### CLASSIFICATION

In [2]:
titanic_train = pd.read_csv("train.csv")
titanic_train.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male  45.0      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

######  Step -1 : Handling Missing Values

In [3]:
titanic_train.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age              0
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         0
dtype: int64

Age, Sex and Fare have no null values.

###### Step -2 Convert Sex into numerical

In [4]:
label_encoder=preprocessing.LabelEncoder()
encoded_sex=label_encoder.fit_transform(titanic_train['Sex'])
titanic_train['Sex']=encoded_sex
titanic_train.head(5)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name  Sex   Age  SibSp  Parch  \
0                            Braund, Mr. Owen Harris    1  22.0      1      0   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...    0  38.0      1      0   
2                             Heikkinen, Miss. Laina    0  26.0      0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)    0  35.0      1      0   
4                           Allen, Mr. William Henry    1  35.0      0      0   

             Ticket     Fare Cabin Embarked  
0         A/5 21171   7.2500   NaN        S  
1          PC 17599  71.2833   C85        C  
2  STON/O2. 3101282   7.9250   NaN        S  
3            113803  53.1000  C123        S  
4            373450   8.0500   NaN        S

###### Step -3 Declaration

In [5]:
tree_model = tree.DecisionTreeClassifier(max_depth=8)

###### Step-4 D.V and I.D.V

In [6]:
features = ["Sex","Age","Fare"]
tree_model.fit(X=titanic_train[features],y=titanic_train['Survived'])

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=8, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=None, splitter='best')

###### Step-5 Opening a file for 'webgraphviz.com'

In [7]:
with open('Dtree.dot','w') as f:
    f=tree.export_graphviz(tree_model,feature_names=["Sex","Age","Fare"],out_file=f)
    

Inferences from Classification:

In [8]:
r=tree.export_text(tree_model,feature_names=features)
print(r)

|--- Sex <= 0.50
|   |--- Fare <= 48.20
|   |   |--- Fare <= 27.82
|   |   |   |--- Fare <= 25.70
|   |   |   |   |--- Fare <= 24.07
|   |   |   |   |   |--- Fare <= 10.48
|   |   |   |   |   |   |--- Fare <= 8.04
|   |   |   |   |   |   |   |--- Fare <= 6.99
|   |   |   |   |   |   |   |   |--- class: 0
|   |   |   |   |   |   |   |--- Fare >  6.99
|   |   |   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |   |--- Fare >  8.04
|   |   |   |   |   |   |   |--- Age <= 19.00
|   |   |   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |   |   |--- Age >  19.00
|   |   |   |   |   |   |   |   |--- class: 0
|   |   |   |   |   |--- Fare >  10.48
|   |   |   |   |   |   |--- Fare <= 14.13
|   |   |   |   |   |   |   |--- Age <= 53.50
|   |   |   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |   |   |--- Age >  53.50
|   |   |   |   |   |   |   |   |--- class: 0
|   |   |   |   |   |   |--- Fare >  14.13
|   |   |   |   |   |   |   |--- Fare <= 15.37
|   |   |   |   |   

Some of the inferences are-

1) Females with fare>48.2, age<8 : Not Survived.

2) Females with fare : (8.04,10.48] and age <=19 : Survived.

3) Males with Age <=6.50 and Fare <=20.83 : Survived.

4) Males with Age <=3.50  and fare : (20.83,23.54]: Not Survived. etc..

#### Prediction

In [9]:
titanic_test = pd.read_csv("test.csv")
titanic_test.head(5)

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Embarked  
0  34.5      0      0   330911   7.8292        Q  
1  47.0      1      0   363272   7.0000        S  
2  62.0      0      0   240276   9.6875        Q  
3  27.0      0      0   315154   8.6625        S  
4  22.0      1      1  3101298  12.2875        S

In [10]:
titanic_test.isna().sum()

PassengerId    0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Embarked       0
dtype: int64

No Null values are present.

In [11]:
label_encoder=preprocessing.LabelEncoder()
encoded_sex=label_encoder.fit_transform(titanic_test['Sex'])
titanic_test['Sex']=encoded_sex
titanic_test.head(5)

PassengerId  Pclass                                          Name  Sex  \
0          892       3                              Kelly, Mr. James    1   
1          893       3              Wilkes, Mrs. James (Ellen Needs)    0   
2          894       2                     Myles, Mr. Thomas Francis    1   
3          895       3                              Wirz, Mr. Albert    1   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)    0   

    Age  SibSp  Parch   Ticket     Fare Embarked  
0  34.5      0      0   330911   7.8292        Q  
1  47.0      1      0   363272   7.0000        S  
2  62.0      0      0   240276   9.6875        Q  
3  27.0      0      0   315154   8.6625        S  
4  22.0      1      1  3101298  12.2875        S

In [12]:
features = ["Sex","Age","Fare"]
test_pred = tree_model.predict(X=titanic_test[features])
predicted_output = pd.DataFrame({"PassengerId":titanic_test['PassengerId'],"Survived":test_pred})
predicted_output.to_csv("Output.csv",index=False)

In [13]:
b=pd.read_csv("Output.csv")

In [14]:
merged = titanic_test.merge(b, on='PassengerId')
merged.to_csv("Prediction.csv", index=False)

In [15]:
ans = pd.read_csv("Prediction.csv")

In [16]:
ans.head(10)

PassengerId  Pclass                                          Name  Sex  \
0          892       3                              Kelly, Mr. James    1   
1          893       3              Wilkes, Mrs. James (Ellen Needs)    0   
2          894       2                     Myles, Mr. Thomas Francis    1   
3          895       3                              Wirz, Mr. Albert    1   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)    0   
5          897       3                    Svensson, Mr. Johan Cervin    1   
6          898       3                          Connolly, Miss. Kate    0   
7          899       2                  Caldwell, Mr. Albert Francis    1   
8          900       3     Abrahim, Mrs. Joseph (Sophie Halaut Easu)    0   
9          901       3                       Davies, Mr. John Samuel    1   

    Age  SibSp  Parch     Ticket     Fare Embarked  Survived  
0  34.5      0      0     330911   7.8292        Q         0  
1  47.0      1      0     363272   7.0000        S         1  
2  62.0      0      0     240276   9.6875        Q         0  
3  27.0      0      0     315154   8.6625        S         0  
4  22.0      1      1    3101298  12.2875        S         1  
5  14.0      0      0       7538   9.2250        S         0  
6  30.0      0      0     330972   7.6292        Q         1  
7  26.0      1      1     248738  29.0000        S         1  
8  18.0      0      0       2657   7.2292        C         1  
9  21.0      2      0  A/4 48871  24.1500        S         0

This is the predicted dataset.